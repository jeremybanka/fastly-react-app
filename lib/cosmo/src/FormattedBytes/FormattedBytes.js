/* @flow */

import * as React from "react";

type Props = {
  value: number,
};

const units = ["B", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"];

export function formatBytes(bytes: number) {
  // handle edge case
  if (bytes <= 0) {
    return `0${units[0]}`;
  }

  let i = Math.floor(Math.log(bytes) / Math.log(1024));

  // keep within bounds of the defined units
  if (i >= units.length) {
    i = units.length - 1;
  }

  const formattedBytes =
    i === 0
      ? (bytes / 1024 ** i).toFixed(0) // no precision needed for bytes
      : (bytes / 1024 ** i).toFixed(1); // default to one precision point

  return `${formattedBytes}${units[i]}`;
}

const FormattedBytes = ({ value, ...rest }: Props): React.Node => (
  <span className="FormattedBytes" {...rest}>
    {formatBytes(value)}
  </span>
);

export default FormattedBytes;
