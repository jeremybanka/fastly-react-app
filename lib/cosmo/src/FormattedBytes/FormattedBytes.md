```jsx
<FormattedBytes value={12345678} />
```
