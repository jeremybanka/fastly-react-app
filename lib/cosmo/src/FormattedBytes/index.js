/* @flow */

import FormattedBytes from "./FormattedBytes";
import { formatBytes } from "./FormattedBytes";

export { formatBytes };

export default FormattedBytes;
