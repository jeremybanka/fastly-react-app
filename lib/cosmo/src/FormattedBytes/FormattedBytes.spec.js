/* @flow */

import React from "react";
import { shallow, render } from "enzyme";

import FormattedBytes from "./FormattedBytes";

describe("FormattedBytes test suite", () => {
  it("renders with className .FormattedBytes", () => {
    const wrapper = shallow(<FormattedBytes value={234523} />);
    expect(wrapper.is(".FormattedBytes")).toBe(true);
  });

  it("correctly handles 0", () => {
    const wrapper = shallow(<FormattedBytes value={0} />);
    expect(wrapper.text()).toBe("0B");
  });

  it("correctly handles negative inputs", () => {
    const wrapper = shallow(<FormattedBytes value={-5} />);
    expect(wrapper.text()).toBe("0B");
  });

  it("correctly handles < 1KB values", () => {
    const wrapper = shallow(<FormattedBytes value={1023} />);
    expect(wrapper.text()).toBe("1023B");
  });

  it("correctly handles KB values", () => {
    const wrapper = shallow(<FormattedBytes value={1024 + 102} />);
    expect(wrapper.text()).toBe("1.1KB");
  });

  it("correctly handles MB values", () => {
    const wrapper = shallow(<FormattedBytes value={1024 ** 2 + 1024 * 100} />);
    expect(wrapper.text()).toBe("1.1MB");
  });

  it("correctly handles unreasonably large inputs", () => {
    const wrapper = shallow(<FormattedBytes value={1024 ** 20} />);
    expect(wrapper.text()).not.toContain("undefined");
  });

  it("renders the stored snapshot", () => {
    const component = <FormattedBytes value={67293902} />;
    expect(render(component)).toMatchSnapshot();
  });
});
