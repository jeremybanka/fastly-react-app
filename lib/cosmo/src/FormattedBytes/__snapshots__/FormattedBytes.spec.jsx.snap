// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormattedBytes test suite renders the stored snapshot 1`] = `
<span
  class="FormattedBytes"
>
  64.2MB
</span>
`;
